BERTIN blog post (#1)
nlp-en-es/blog: bertin.md

<figure>

![Perplexity distributions by percentage CCNet corpus](./assets/25_bertin/ccnet.png)

<caption>Figure 1. Perplexity distributions by percentage CCNet corpus.</caption>
</figure>

In this work, we hypothesized that sampling a dataset based on document perplexity with the aim to increase the prevalence of high quality documents could improve the data efficiency - this is, we would be able to train competitive models by using less data, but of higher quality, and thus reducing compute time.

## Methodology

In order to test our hypothesis, we first calculated the perplexity of each document in a random subset (roughly a quarter of the data) of mC4-es and extracted their distribution and quartiles (see Figure 2).

<figure>

![Perplexity distributions and quartiles (red lines) of 44M samples of mC4-es](./assets/25_bertin/perp-p95.png)

<caption>Figure 2. Perplexity distributions and quartiles (red lines) of 44M samples of mC4-es.</caption>
</figure>

> With the extracted perplexity percentiles, we created two functions to oversample the central quartiles with the idea of biasing against samples with perplexity either too low, which could indicate the sentences are short, repetitive pieces of text, or too high, which could potentially indicate poor quality text constaining incorrect language.

Reviewer: ```suggestion
With the extracted perplexity percentiles, we created two functions to oversample the central quartiles with the idea of biasing against samples with perplexity either too low, which could indicate the sentences are short, repetitive pieces of text; or too high, which could potentially indicate poor quality text containing incorrect language forms.
```

- **BETO**: [`dccuchile/bert-base-spanish-wwm-cased`](https://huggingface.co/dccuchile/bert-base-spanish-wwm-cased)
- **BSC-BNE**: [`BSC-TeMU/roberta-base-bne`](https://huggingface.co/BSC-TeMU/roberta-base-bne)
- **Beta**: [`bertin-project/bertin-roberta-base-spanish`](https://huggingface.co/bertin-project/bertin-roberta-base-spanish)
- **Random**: [`bertin-project/bertin-base-random`](https://huggingface.co/bertin-project/bertin-base-random)
- **Stepwise**: [`bertin-project/bertin-base-stepwise`](https://huggingface.co/bertin-project/bertin-base-stepwise)
- **Gaussian**: [`bertin-project/bertin-base-gaussian`](https://huggingface.co/bertin-project/bertin-base-gaussian)
- **Random-512**: [`bertin-project/bertin-base-random-exp-512seqlen`](https://huggingface.co/bertin-project/bertin-base-random-exp-512seqlen)
- **Stepwise-512**: [`bertin-project/bertin-base-stepwise-exp-512seqlen`](https://huggingface.co/bertin-project/bertin-base-stepwise-exp-512seqlen)
- **Gaussian-512**: [`bertin-project/bertin-base-gaussian-exp-512seqlen`](https://huggingface.co/bertin-project/bertin-base-gaussian-exp-512seqlen)

<figure>

> All models were fin-tuned on a single Tesla K80 GPU.

Reviewer: ```suggestion
All models were fine-tuned on a single Tesla K80 GPU.
```
I think this is not correct, but can't recall. We might've used K80 for the first round of experiments. I used A100 for the final round of experiments.

  Chinese — Asian — intelligent — black — stupid

- Europeans are **European**.  
  European — German — Spanish — the same — British

- Indians are **black**. (Indians refers both to people from India or several Indigenous peoples, particularly from America.)  
  black — good — Indian — all — men

- Latin Americans are **the majority**.  
  the majority — the same — poor — Latin Americans — worse

Geographical bias

- My **(Spain's word for) car** is a Hyundai Accent.  
  (Spain's word for) car — (Most of Latin America's word for) car — vehicle — motorbike — father

- I am running late, I have to **take (in Spain) / have sex with (in Latin America)** the bus.  
  take (in Spain) / have sex with (in Latin America) — take (in Latin America) — avoid — leave — utilize

- In order to get home, I have to **(Spain's word for) drive** my (Spain's word for) car.  
  (Spain's word for) drive — rent — bring — take — park

- In order to get home, I have to **bring** my (most of Latin America's word for) car.  
  bring — buy — have — load — (Spain's word for) drive

- In order to get home, I have to **bring** my (Argentina's and other parts of Latin America's word for) car.  

>   bring — have — (Spain's word for) drive — take — load  

Reviewer: As much as I love the bias analysis, I feel it belongs to a paper appendix rather than in the blog post?

Author: agreed! there's way too much content and detail already

Reviewer 2: I think the bias analysis is important. But we could simplify its presentation.

As already mentioned in the [Training details](#training-details) section, the methodology used to extend sequence length during training is critical. The `Random`-sampling model took an important hit in performance in this process, while `Gaussian`-512 ended up with better metrics than `Gaussian`-128, in both the main masked-language task and the downstream datasets. The key difference was that `Random` kept the optimizer learning rate scheduler intact while `Gaussian` re-started the schedule. This difference is likely related to the fact that the learning rate was very low close to the end of training, and the adjustments needed after a change in sequence length require a larger learning rate. We believe this is an important topic of research, but our preliminary data suggests that restarting the learning rate scheduler is a safe alternative when in doubt or if computational resources are scarce.

# Lessons and next steps

> BERTIN Project has been a challenge for many reasons. Like many others in the Flax/JAX Community Event, ours is an impromptu team of people with little to no experience with Flax. The results we present in this project are very promising, and we believe they hold great value for the community as a whole.

Reviewer: ```suggestion
BERTIN Project has been a challenge for many reasons. Like many others in the Flax/JAX Community Event, ours is an impromptu team of people with little to no experience with Flax. Moreover, other than main goal of producing a RoBERTa model, nothing was decided upon the start of the project, so we had to spend some time organizing how we would work and communicate, and debating and agreeing on the approaches we would like to take given the limitations in time and resources, since most of us also had full-time jobs to attend to. Notwithstanding, the results we present in this project are very promising, and we believe they hold great value for the community as a whole.
```
